### PR TITLE
Bail early on initdb logic if we detect that it isn't needed

### PR DESCRIPTION
**Pull request by tianon**

### CODE REVIEW

**Comment by tianon** [comment]
This probably conflicts with https://github.com/docker-library/mongo/pull/182, but I'll happily rebase either PR if so.

File: 3.0/docker-entrypoint.sh [hunk 1]
```diff
@@ -108,21 +109,46 @@ if [ "$originalArgOne" = 'mongod' ]; then
 		set -- "${mongodHackedArgs[@]}"
 	fi
 
+	# pre-check a few factors to see if it's even worth bothering with initdb
+	shouldPerformInitdb=
+	if [ -z "$shouldPerformInitdb" ]; then
+		# if we've got any MONGO_INITDB_xxx environment variables set, we should initdb
+		for var in "${!MONGO_INITDB_@}"; do
```
Comment (yosifkit) [inline 3.0/docker-entrypoint.sh]: Maybe this should be more specific about needing user and pass together and warn or even fail if they only have one?  Technically `MONGO_INITDB_DATABASE` doesn't matter unless they have scripts since a database is not created unless something is inserted into a collection.

Comment (tianon) [inline 3.0/docker-entrypoint.sh]: I think https://github.com/docker-library/mongo/blob/8b9ff681751fbb4fbeaf028ad9420014303755b9/3.5/docker-entrypoint.sh#L118 or https://github.com/docker-library/mongo/blob/8b9ff681751fbb4fbeaf028ad9420014303755b9/3.5/docker-entrypoint.sh#L196 would be more appropriate places to have that check/warning (probably the former so it happens ASAP and before any `initdb` output).

Comment (yosifkit) [inline 3.0/docker-entrypoint.sh]: So perhaps the loop could be simplified to just be part of the `if` above it:
```diff
 	file_env 'MONGO_INITDB_ROOT_USERNAME'
 	file_env 'MONGO_INITDB_ROOT_PASSWORD'
+	# pre-check a few factors to see if it's even worth bothering with initdb
+	shouldPerformInitdb=
 	if [ "$MONGO_INITDB_ROOT_USERNAME" ] && [ "$MONGO_INITDB_ROOT_PASSWORD" ]; then
 		# if we have a username/password, let's set "--auth"
 		_mongod_hack_ensure_arg '--auth' "$@"
  		set -- "${mongodHackedArgs[@]}"
+		shouldPerformInitdb="true" # 1/yes/true/something that works for -n and not -z
+	elif [ "$MONGO_INITDB_ROOT_USERNAME" ] || [ "$MONGO_INITDB_ROOT_PASSWORD" ]; then
+		have="${MONGO_INITDB_ROOT_USERNAME:+MONGO_INITDB_ROOT_USERNAME}${MONGO_INITDB_ROOT_PASSWORD:+MONGO_INITDB_ROOT_PASSWORD}"
+		haveNot="${MONGO_INITDB_ROOT_USERNAME:+MONGO_INITDB_ROOT_PASSWORD}${MONGO_INITDB_ROOT_PASSWORD:+MONGO_INITDB_ROOT_USERNAME}"
+		echo >&2 "'$have' specified but '$haveNot' is missing"
+		echo >&2 "both must be specified for a user to be created"
+		exit 1
  	fi
```

Comment (tianon) [inline 3.0/docker-entrypoint.sh]: Ok, feel free to commit your changes to the branch. :wink:

Comment (tianon) [inline 3.0/docker-entrypoint.sh]: (Although IMO that `have` and `haveNot` is pretty complicated -- even as a seasoned shell-er, this is a bit hard to parse, so I'd rather either have another `if` or just make the error message generic and make the user figure out which one they're missing)

**Comment by yosifkit** [comment]
New commit added!

```console
$ docker run -it --rm -e MONGO_INITDB_ROOT_USERNAME=admin 3925e6d4ebf3

error: missing 'MONGO_INITDB_ROOT_USERNAME' or 'MONGO_INITDB_ROOT_PASSWORD'
       both must be specified for a user to be created.

$ echo $?
1
```

**Comment by tianon** [comment]
LGTM
